src/cfnlint/rules/resources/codepipeline/CodepipelineStageActions.py

```diff
@@ -0,0 +1,254 @@
+                }
+            },
+        },
+    }
+
+    KEY_MAP = {
+        'InputArtifacts': 'InputArtifactRange',
+        'OutputArtifacts': 'OutputArtifactRange',
+    }
+
+    def check_artifact_counts(self, action, artifact_type, path):
+        """Check that artifact counts are within valid ranges."""
+        matches = []
+
+        action_type_id = action.get('ActionTypeId')
+        owner = action_type_id.get('Owner')
+        category = action_type_id.get('Category')
+        provider = action_type_id.get('Provider')
+
+        constraints = self.CONSTRAINTS.get(owner, {}).get(category, {}).get(provider, {})
+        if not constraints:
+            return matches
+        artifact_count = len(action.get(artifact_type, []))
+
+        constraint_key = self.KEY_MAP[artifact_type]
+        if isinstance(constraints[constraint_key], tuple):
+            min_, max_ = constraints[constraint_key]
+            if not (min_ <= artifact_count <= max_):
+                message = (
+                    'Action "{action}" declares {number} {artifact_type} which is not in '
+                    'expected range [{a}, {b}].'
+                ).format(
+                    action=action['Name'],
+                    number=artifact_count,
+                    artifact_type=artifact_type,
+                    a=min_,
+                    b=max_
+                )
+                matches.append(RuleMatch(
+                    path + [artifact_type],
+                    message
+                ))
+        else:
+            if artifact_count != constraints[constraint_key]:
+                message = (
+                    'Action "{action}" declares {number} {artifact_type} which is not the '
+                    'expected number [{a}].'
+                ).format(
+                    action=action['Name'],
+                    number=artifact_count,
+                    artifact_type=artifact_type,
+                    a=constraints[constraint_key]
+                )
+                matches.append(RuleMatch(
+                    path + [artifact_type],
+                    message
+                ))
+
+        return matches
+
+    def check_action_keys(self, action, path):
+        """Check that action has required keys inside."""
+        matches = []
+
+        if 'Name' not in action:
+            message = 'Expecting Name key in Action.'
+            matches.append(RuleMatch(
+                path,
+                message
+            ))
+        if 'ActionTypeId' not in action:
+            message = 'Expecting ActionTypeId key in Action.'
+            matches.append(RuleMatch(
+                path,
+                message
+            ))
+
+        return matches
+
+    def check_owner(self, action, path):
+        """Check that action type owner is valid."""
+        matches = []
+
+        owner = action.get('ActionTypeId').get('Owner')
+        if owner not in self.VALID_OWNER_STRINGS:
```

Comment:
I'm debating if this should be.
`if owner not in self.VALID_OWNER_STRINGS and owner is not None:` since the Owner check will be covered by the required properties check (E3003) if someone misses Owner with this version they will get double errors.

```
E3003 Property Owner missing from resource DeployPipeline
test/templates/quickstart/codepipeline.yaml:441:11

E2541 For all currently supported action types, the only valid owner strings are AWS, Custom, ThirdParty
test/templates/quickstart/codepipeline.yaml:444:15
```

where as with the fix it would just be 
```
E3003 Property Owner missing from resource DeployPipeline
test/templates/quickstart/codepipeline.yaml:441:11
```